components/merkledb/examples/migration_with_helper.rs

```diff
@@ -0,0 +1,153 @@
+    ObjectHash,
+};
+
+use migration::{
+    check_data_after_flush, check_data_after_merge, check_data_before_flush, create_initial_data,
+    v1, v2,
+};
+
+mod migration;
+
+fn create_migration(helper: &mut MigrationHelper) {
+    println!("\nStarted migration");
+
+    {
+        let new_data = helper.new_data();
+        let old_data = helper.old_data();
+
+        let old_schema = v1::Schema::new(old_data);
+        let mut new_schema = v2::Schema::new(new_data);
+
+        // Move `ticker` and `divisibility` to `config`.
+        let config = v2::Config {
+            ticker: old_schema.ticker.get().unwrap(),
+            divisibility: old_schema.divisibility.get().unwrap_or(0),
+        };
+        new_schema.config.set(config);
+        // Mark these two indexes for removal.
+        new_data.create_tombstone("ticker");
+        new_data.create_tombstone("divisibility");
+    }
+
+    // Migrate wallets using schema:
+    // `Wallet::public_key` field will be removed.
+    // `Wallet::history_hash` field will be added.
+    // Wallets and history from username Eve will be removed.
+    migrate_wallets(&helper);
+}
+
+/// Provides migration of wallets with schema and `MigrationHelper`.
+fn migrate_wallets(helper: &MigrationHelper) {
+    let old_schema = v1::Schema::new(helper.old_data());
+    let mut new_schema = v2::Schema::new(helper.new_data());
+
+    // Migrate wallets.
+    for (i, (public_key, wallet)) in old_schema.wallets.iter().enumerate() {
+        if wallet.username == "Eve" {
+            // We don't like Eves 'round these parts. Remove her transaction history
+            // and don't migrate the wallet.
+            helper
+                .new_data()
+                .create_tombstone(("histories", &public_key));
+        } else {
+            // Merkelize the wallet history.
+            let mut history = new_schema.histories.get(&public_key);
+            history.extend(&old_schema.histories.get(&public_key));
+
+            let new_wallet = v2::Wallet {
+                username: wallet.username,
+                balance: wallet.balance,
+                history_hash: history.object_hash(),
+            };
+            new_schema.wallets.put(&public_key, new_wallet);
+        }
+
+        if i % 1_000 == 999 {
+            println!("Processed {} wallets", i + 1);
+        }
+    }
+}
+
+fn main() {
+    // Creating a temporary DB and filling it with some data.
+    let db = create_initial_data();
+
+    // Creating helper to perform migration.
+    let mut helper = MigrationHelper::new(db.clone(), "test");
+
+    {
+        // State before migration.
+        let old_schema = v1::Schema::new(helper.old_data());
+        println!("Before migration:");
+        old_schema.print_wallets();
+    }
+
+    // Migrate the data.
+    create_migration(&mut helper);
+
+    // Call `MigrationHelper::finish` to merge changes to the database.
+    if let Err(err) = helper.finish() {
```

Comment:
I'd use `helper.finish().expect("Migration finish failed")` to get a non-zero exit code on error.